ni_measurementlink_service/_internal/parameter/serialization_strategy.py

```diff
@@ -1,257 +1,212 @@
-    """Abstract Specific Encoder(Callable) as Vector Encoder Callable that takes in field index.
-
-    Args
-    ----
-       encoder (Callable[[int, bool, bool], Callable]): Specific encoder(Callable) that takes in
-       field_index, is_repeated, is_packed and returns the Low-level Encode Callable.
+def _vector_encoder(encoder: EncoderConstructor, is_packed=True) -> PartialEncoderConstructor:
+    """Constructs a vector (array) encoder constructor.
 
-       is_packed (bool, optional): Represents if the encoder supports packed type. Defaults to True.
-
-    Returns
-    -------
-        Callable[[int],Callable]: Callable Encoder for 1D Array types that takes in
-        field_index and returns the Low-level Encode Callable.
+    Takes a field index and returns an Encoder.
 
+    This class returns the Encoder with is_repeated set to True
+    and is_packed defaulting to True.
     """
 
-    def vector_encoder(field_index):
+    def vector_encoder(field_index: int) -> Encoder:
         is_repeated = True
         return encoder(field_index, is_repeated, is_packed)
 
     return vector_encoder
 
 
-def _scalar_decoder(decoder) -> Callable[[int, str], Callable]:
-    """Abstract Specific Decoder(Callable) as Scalar Decoder Callable that takes in field index,key.
-
-    Args
-    ----
-        decoder (Callable[[int, bool, bool], Callable]): Specific decoder(Callable) that takes in
-        field_index, is_repeated, is_packed,  key, new_default and
-        returns the Low-level Decode Callable.
+def _scalar_decoder(decoder: DecoderConstructor) -> PartialDecoderConstructor:
+    """Constructs a scalar decoder constructor.
 
-    Returns
-    -------
-        Callable[[int,str],Callable]: Callable Decoder for scalar types that takes in
-        field_index, key and returns the Low-level Decode Callable.
+    Takes a field index and a key and returns a Decoder.
 
+    This class returns the Decoder with is_repeated set to False
+    and is_packed set to False.
     """
 
-    def scalar_decoder(field_index, key):
+    def _unsupported_new_default(message: Message) -> Message:
+        raise NotImplementedError(
+            "This function should not be called. Verify that you are using up-to-date and compatible versions of the ni-measurementlink-service and protobuf packages."
+        )
+
+    def scalar_decoder(field_index: int, key: Key) -> Decoder:
         is_repeated = False
         is_packed = False
-        return decoder(field_index, is_repeated, is_packed, key, None)
+        return decoder(field_index, is_repeated, is_packed, key, _unsupported_new_default)
 
     return scalar_decoder
 
 
-def _vector_decoder(decoder, is_packed=True) -> Callable[[int, str], Callable]:
-    """Abstract Specific Decoder(Callable) as Vector Decoder Callable that takes in field index,key.
+def _vector_decoder(decoder: DecoderConstructor, is_packed=True) -> PartialDecoderConstructor:
```

Comment:
Using a TypeVar in a TypeAlias such as DecoderConstructor has consequences that I didn't consider: it makes the alias generic. If you use DecoderConstructor without specifying a type variable, it's equivalent to DecoderConstructor[Any].

Here's what reveal_type(_vector_decoder) says: 
```
ni_measurementlink_service\_internal\parameter\serialization_strategy.py:105: note: Revealed type is "def (decoder: def (builtins.int, builtins.bool, builtins.bool, Any, def (google.protobuf.message.Message) -> google.protobuf.message.Message) -> def (builtins.memoryview, builtins.int, builtins.int, google.protobuf.message.Message, builtins.dict[Any, Any]) -> builtins.int, is_packed: Any =) -> def (builtins.int, Any) -> def (builtins.memoryview, builtins.int, builtins.int, google.protobuf.message.Message, builtins.dict[Any, Any]) -> builtins.int"
```

One solution is to add `[Key]` to all references to these type aliases, including the references to Decoder inside the other type aliases:
```
PartialDecoderConstructor: TypeAlias = Callable[[int, Key], Decoder[Key]]
DecoderConstructor: TypeAlias = Callable[[int, bool, bool, Key, NewDefault], Decoder[Key]]
```

However, this is really starting to diverge from the (corrected) types in typeshed, so maybe we're better off specifying `Key = FieldDescriptor` and casting whenever we pass in an `int`.

Reply:
> How do you cast an int to a FieldDescriptor? They're very different things.

```python
        inner_decoder = decoder(field_index, cast(FieldDescriptor, field_index))
        ...
        position = inner_decoder(
            parameter_bytes_memory_view,
            position + encoder._TagSize(field_index),  # type: ignore[attr-defined]
            len(parameter_bytes),
            cast(Message, None),  # unused - See serialization_strategy._vector_decoder._new_default
            cast(Dict[FieldDescriptor, Any], overlapping_parameters_by_id),
        )
```

This has no effect at run time, but it lies to the type checker, saying "trust me, this is a FieldDescriptor". 

> Also you're talking about going back to a TypeAlias like this?
>
> Key: TypeAlias = FieldDescriptor

Yes.

> And does that mean this will return Dict[FieldDescriptor, Any]? As well as many other methods.
>
> ```
>def deserialize_parameters(
>    parameter_metadata_dict: Dict[int, ParameterMetadata], parameter_bytes: bytes
>) -> Dict[int, Any]:
> ```

No. Try to use int in the ni-measurementlink-python code. Cast to FieldDescriptor when directly interfacing with protobuf. I would be surprised if any code outside of serialization_strategy.py and serializer.py needs to use the FieldDescriptor version of these types.